### PR TITLE
Support clipBehavior changes in hot reload

### LINKED ISSUE #30863: Changes to PhysicalModel's clipBehavior not reflected by hot reload
I have this code in an otherwise empty app:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('Title!'),
        ),
        body: Center(
          child: PhysicalModel(
            elevation: 2.0,
            borderRadius: const BorderRadius.all(Radius.circular(10.0)),
            shape: BoxShape.rectangle,
            color: Colors.grey,
            child: Container(
              height: 100,
              width: 100,
              color: Colors.blue,
            ),
          ),
        ),
      ),
    );
  }
}
```

If I run the app, the corners of the blue container are not clipped, since PhysicalModel's clipBehavior defaults to `none`. If I try to remedy this by adding `clipBehavior: Clip.antiAlias` to the PhysicalModel, nothing happens when I hot reload. I have to do a full hot restart to see the change.

PhysicalModel's [updateRenderObject method](https://github.com/flutter/flutter/blob/ab06d033ca210964c541f09c53803590aa95e865/packages/flutter/lib/src/widgets/basic.dart#L884) doesn't update the clipBehavior:

```dart
  @override
  void updateRenderObject(BuildContext context, RenderPhysicalModel renderObject) {
    renderObject
      ..shape = shape
      ..borderRadius = borderRadius
      ..elevation = elevation
      ..color = color
      ..shadowColor = shadowColor;
  }
```

That's the reason a hot restart is required. Is this by design? This would mean it's impossible to rebuild the PhysicalModel with a different clipBehavior under normal circumstances as well:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  int count = 0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('Title!'),
        ),
        body: Center(
          child: PhysicalModel(
            elevation: 2.0,
            borderRadius: const BorderRadius.all(Radius.circular(10.0)),
            shape: BoxShape.rectangle,
            color: Colors.grey,
            clipBehavior: count % 2 == 0 ? Clip.none : Clip.antiAlias,
            child: GestureDetector(
              onTap: () {
                setState(() => count++);
              },
              child: Container(
                height: 100,
                width: 100,
                color: Colors.blue,
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

## Logs

Flutter doctor: 

```
[✓] Flutter (Channel stable, v1.2.1, on Mac OS X 10.13.6 17G5019, locale en-US)
    • Flutter version 1.2.1 at /Users/redbrogdon/Library/flutter
    • Framework revision 8661d8aecd (8 weeks ago), 2019-02-14 19:19:53 -0800
    • Engine revision 3757390fa4
    • Dart version 2.1.2 (build 2.1.2-dev.0.0 0a7dcf17eb)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/redbrogdon/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = /Users/redbrogdon/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • ios-deploy 1.9.4
    • CocoaPods version 1.6.1

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 27.1.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] IntelliJ IDEA Community Edition (version 2018.3.5)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 33.4.2
    • Dart plugin version 183.5912.23

[✓] Connected device (1 available)
    • iPhone XR • 2998CF6D-5051-46A1-B4FC-5336EDDD73E2 • ios • iOS 12.1 (simulator)

• No issues found!
```

### PR DESCRIPTION
**Pull request by liyuqian**
## Description

Make `_RenderCustomClip`'s `clipBehavior` non-final so we can update it during `updateRenderObject`. This will support `clipBehavior` changes in hot reload.

## Related Issues

Fixes #30863

## Tests

I added the following tests:

* ClipRect updates clipBehavior in updateRenderObject
* ClipRRect updates clipBehavior in updateRenderObject
* ClipOval updates clipBehavior in updateRenderObject
* ClipPath updates clipBehavior in updateRenderObject
* PhysicalModel updates clipBehavior in updateRenderObject
* PhysicalShape updates clipBehavior in updateRenderObject

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 1]
```diff
@@ -762,3 +762,14 @@ class _TestTicker extends Ticker {
     super.dispose();
   }
 }
+
+mixin UpdateCount on RenderObjectWidget {
```
Comment (xster) [inline packages/flutter_test/lib/src/widget_tester.dart]: I wouldn't add this new mixin to widget_tester. Just checking that the renderObject's property changed as result of pumping a new widget is ok.

**Review by xster — COMMENTED** [review]
LGTM

File: packages/flutter_test/lib/src/widget_tester.dart (hunk 1 above)
Comment (liyuqian) [reply packages/flutter_test/lib/src/widget_tester.dart]: Done :)

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**